packages/datadog-plugin-google-cloud-pubsub/src/pubsub-push-subscription.js

```diff
@@ -0,0 +1,193 @@
+const id = require('../../dd-trace/src/id')
+const log = require('../../dd-trace/src/log')
+
+class GoogleCloudPubsubPushSubscriptionPlugin extends TracingPlugin {
+  static get id () { return 'google-cloud-pubsub-push-subscription' }
+
+  constructor (...args) {
+    super(...args)
+
+    // Use addSub to subscribe to HTTP server request start events
+    this.addSub('apm:http:server:request:start', (ctx) => {
+      this.#handlePubSubRequest(ctx)
+    })
+  }
+
+  #handlePubSubRequest (ctx) {
+    const { req } = ctx
+    const userAgent = req.headers['user-agent'] || ''
+    if (req.method !== 'POST' || !userAgent.includes('APIs-Google')) {
+      return false
+    }
+
+    if (req.headers['x-goog-pubsub-message-id']) {
+      this.#createDeliverySpanAndActivate(ctx)
+      return true
+    }
+
+    log.warn(
+      '[PubSub] No x-goog-pubsub-* headers detected. pubsub.delivery spans will not be created. ' +
+      'Add --push-no-wrapper-write-metadata to your subscription.'
+    )
+    return false
+  }
+
+  #createDeliverySpanAndActivate (ctx) {
+    const { req, res } = ctx
+    const messageData = this.#parseMessage(req)
+    if (!messageData) {
+      return
+    }
+
+    const originalContext = this.#extractContext(messageData)
+    const pubsubRequestContext = this.#reconstructPubSubContext(messageData.attrs) || originalContext
+    const isSameTrace = pubsubRequestContext &&
+      originalContext?.toTraceId() === pubsubRequestContext.toTraceId()
+
+    // Create and enter the delivery span so it becomes the active span
+    const deliverySpan = this.#createDeliverySpan(
+      messageData,
+      isSameTrace ? pubsubRequestContext : originalContext,
+      isSameTrace ? null : pubsubRequestContext
+    )
+
+    if (!deliverySpan) {
+      return
+    }
+
+    // Activate the delivery span so the HTTP server span will be a child of it
+    this.enter(deliverySpan, { req, res })
+
+    // Finish span when HTTP response lifecycle completes
+    this.#attachFinishHandlers(deliverySpan, res)
+  }
+
+  #attachFinishHandlers (span, res) {
+    let finished = false
+    const finishOnce = (error) => {
+      if (finished) return
+      finished = true
+      if (error) {
+        span.setTag('error', error)
+      }
+      span.finish()
+    }
+
+    res.once('finish', () => finishOnce())
+    res.once('close', () => finishOnce())
+    res.once('error', (err) => finishOnce(err))
+  }
+
+  #parseMessage (req) {
+    const subscription = req.headers['x-goog-pubsub-subscription-name']
+    const message = {
+      messageId: req.headers['x-goog-pubsub-message-id'],
+      publishTime: req.headers['x-goog-pubsub-publish-time']
+    }
+
+    const topicName = req.headers['pubsub.topic'] || 'push-subscription-topic'
+    return { message, subscription, attrs: req.headers, topicName }
+  }
+
+  #extractContext (messageData) {
+    return this.tracer.extract('text_map', messageData.attrs) ?? undefined
```

Comment:
Is the `?? undefined` needed? It would otherwise return potentially `null`, while that seems fine as far as I can tell from the call sites.